### PR TITLE
[SYCL] Move SYCLLowerWGLocalMemoryPass to OptimizerEarlyEPCallback

### PR DESCRIPTION
**Pull request by jsji**
The pass transforms __sycl_allocateLocalMemory call to access of global
variable @WGLocalMem .

Move the transform to beginning of the optimizer since the access
could enable more optimization than the function call.

### CODE REVIEW

**Comment by jsji** [comment]
Similar to https://github.com/intel/llvm/pull/16183 but tweak the position of the pass to satisfy the requirements.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 1]
```diff
@@ -1043,7 +1043,14 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
             /*ExcludeAspects=*/{"fp64"}));
         MPM.addPass(SYCLPropagateJointMatrixUsagePass());
       });
-    else if (LangOpts.SYCLIsHost && !LangOpts.SYCLESIMDBuildHostCode)
+      PB.registerOptimizerEarlyEPCallback([](ModulePassManager &MPM,
+                                             OptimizationLevel Level,
+                                             ThinOrFullLTOPhase) {
+        // Allocate static local memory in SYCL kernel scope for each allocation
+        // call.
```
Comment (Naghasan) [inline clang/lib/CodeGen/BackendUtil.cpp:1050]:
```suggestion
        // Lowers __sycl_allocateLocalMemory  and __sycl_dynamicLocalMemoryPlaceholder
        // builtin calls.
```

Nit, the pass also handles `work_group_scratch_memory` (https://github.com/intel/llvm/pull/15061) but I didn't change this comment as I didn't touch this part.

**Comment by bader** [comment]
@jsji, is this PR ready for review?

Thanks for putting the motivation for the change to the description. It sounds reasonable except this part:

> In addition, intel gpu compiler has a pass to transform global variable
> in addrspace(3) to alloca, we must run SYCLLowerWGLocalMemoryPass before
> it.

DPC++ runs all passes before the translation to SPIR-V and IGC runs its passes after the translation to SPIR-V, so existing location satisfies the requirement already.

**Review by elizabethandrews — COMMENTED** [review]
FE change looks ok to me but I do not know the functionality well enough to approve the new pass location. I defer to @bader for approval

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 2]
```diff
@@ -1043,7 +1043,14 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
             /*ExcludeAspects=*/{"fp64"}));
         MPM.addPass(SYCLPropagateJointMatrixUsagePass());
       });
-    else if (LangOpts.SYCLIsHost && !LangOpts.SYCLESIMDBuildHostCode)
+      PB.registerOptimizerEarlyEPCallback([](ModulePassManager &MPM,
```
Comment (bader) [inline clang/lib/CodeGen/BackendUtil.cpp:1046]: What is the reason to use `registerOptimizerEarlyEPCallback` instead of `registerPipelineStartEPCallback`?

Comment (jsji) [reply clang/lib/CodeGen/BackendUtil.cpp:1046]: As mentioned in https://github.com/intel/llvm/pull/16183/files/43526397fa2b4c559666fdfd6f302bfba7409255#r1857546652, we need to make sure this pass run after *AlwaysIniner*, moving to` PipelineStart` is what was tried in https://github.com/intel/llvm/pull/16183, it is before alwaysinliner and too early.

Comment (bader) [reply clang/lib/CodeGen/BackendUtil.cpp:1046]: @jsji, could you please extend the comment at lines 1049-1050 with the requirement to run _AlwaysInliner_ pass before _SYCLLowerWGLocalMemoryPass_?

**Review by bader — APPROVED** [review]

File: clang/lib/CodeGen/BackendUtil.cpp (hunk 2 above)
Comment (jsji) [reply clang/lib/CodeGen/BackendUtil.cpp:1046]: For sure.

**Comment by wenju-he** [comment]
If we are sure that the pass needs to be moved to an earlier stage, I had an idea that we can remove the pass's implicit dependency on alwaysinliner pass and move the pass to pipeline start: https://github.com/wenju-he/llvm/commit/e6ce5846246f3fa0f2e0b4e07a154ec11d6d15d9
Please review if it makes sence.

Moving the pass to pipestart has a benefit that the lowering from global variable to alloca could also be done earlier in the pipeline, which might be helpful for the performance of OpenCL which doesn't have LTO.

**Comment by jsji** [comment]
> If we are sure that the pass needs to be moved to an earlier stage, I had an idea that we can remove the pass's implicit dependency on alwaysinliner pass and move the pass to pipeline start: [wenju-he@e6ce584](https://github.com/wenju-he/llvm/commit/e6ce5846246f3fa0f2e0b4e07a154ec11d6d15d9) Please review if it makes sence.

Thanks @wenju-he ! Yes, looks good idea to handle the inlining of these two functions inside the pass itself. Please post your PR for review instead.

**Comment by wenju-he** [comment]
> Please post your PR for review instead.

https://github.com/intel/llvm/pull/16356

**Comment by jsji** [comment]
Close, use #16356 instead.

**Pull request closed without merging**